pkg/reconciler/taskrun/taskrun.go

```diff
@@ -275,8 +284,15 @@ func (c *Reconciler) reconcile(ctx context.Context, tr *v1alpha1.TaskRun) error
 	}
 	cloudevent.InitializeCloudEvents(tr, prs)
 
+	return taskSpec, rtr, nil
+}
+
+func (c *Reconciler) reconcile(ctx context.Context, tr *v1alpha1.TaskRun,
```

Comment:
Should these be *v1beta1 somehow? After `prepare` they should be converted to beta?

Also I think I should move the emit event bit out of `reconcile` to `Reconcile`, so that all events are generated there.

Reply:
Nope. The are converted to `beta` and then to `alpha` to make sure they are in `beta` shape — but the internal (and stored) representation is still `v1alpha1`. But the `v1alpha1` struct you have after `prepare` are **sure** to be in the `v1beta1` shape, so you can/should/have-to use *only* `v1beta1` fields :angel: 

I know it's a bit confusing right now. This will change with #2413 and #2410 :angel:

Reply:
ack, thanks

Reply:
agree about event emitting being centralized into one area of the file 👍 

Also, could we add a comment about the conditions under which an error _should_ be returned by `func reconcile(...)`?  So often I see errors caught in the code but we `return nil` regardless.  I've never totally understood the logic of when to return an error from reconcile and when not to.

Reply:
I'm trying to make the logic about errors and events more consistent in the controller, so I will add some comments about this. 
One thing that I'd like to do eventually is also to start using `PermanentError`s, from knative/pkg. When `Reconcile` returns a `PermanentError` the key is not re-queued.